src/frequenz/sdk/_internal/_logging.py

```diff
@@ -0,0 +1,300 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Logging utilities for the SDK."""
+
+import logging
+from collections.abc import Mapping
+from datetime import datetime, timedelta
+from types import TracebackType
+
+_ExcInfoType = (
+    bool
+    | BaseException
+    | tuple[None, None, None]
+    | tuple[type[BaseException], BaseException, TracebackType | None]
+    | None
+)
+
+DEFAULT_RATE_LIMIT = timedelta(minutes=15)
+
+# The standard logging.py file uses variadic arguments in the logging methods, but the
+# type hints file has more specific parameters.  pylint is not able to handle this, so
+# we need to suppress the warning.
+#
+# pylint: disable=arguments-differ
+
+
+class RateLimitedLogger:
+    """Logger that limits the rate of logging messages.
+
+    The first message is logged immediately.  Subsequent messages are ignored until the
+    rate limit interval has elapsed.  After that the next request goes through, and so on.
+
+    This allows a new outage to be reported immediately and subsequent logs to be
+    rate-limited.
+
+    When an outage has been resolved, the `reset()` method may be used to reset the
+    logger and the next message will get logged immediately.
+    """
+
+    def __init__(
+        self,
+        logger: logging.Logger,
+        rate_limit: timedelta = DEFAULT_RATE_LIMIT,
+    ) -> None:
+        """Initialize the logger.
+
+        Args:
+            logger: Logger to rate-limit.
+            rate_limit: Time interval between two log messages.
+        """
+        self._logger = logger
+        self._started: bool = False
+        self._last_log_time: datetime | None = None
+        self._rate_limit: timedelta = rate_limit
+
+    def set_rate_limit(self, rate_limit: timedelta) -> None:
+        """Set the rate limit for the logger.
+
+        Args:
+            rate_limit: Time interval between two log messages.
+        """
+        self._rate_limit = rate_limit
```

Comment:
Why not just making it public?

```suggestion
        self.rate_limit: timedelta = rate_limit
        """The rate limit for the logger."""
```